src/camera.rs

```diff
@@ -0,0 +1,162 @@
+            polar_angle: clamp(
+                polar_angle,
+                options.polar_angle_distance_min,
+                f32::consts::PI - options.polar_angle_distance_min,
+            ),
+            origin: Point3::origin(),
+            up: Vector3::z(),
+            options,
+        }
+    }
+
+    pub fn set_screen_size(&mut self, screen_size: [f32; 2]) {
+        self.aspect_ratio = screen_size[0] / screen_size[1];
+    }
+
+    pub fn reset_origin(&mut self) {
+        self.origin = Point3::origin();
+    }
+
+    pub fn pan_ground(&mut self, dx: f32, dy: f32) {
+        let pan_factor = self.options.speed_pan * self.radius / self.options.radius_max;
+        let ground_translation = Vector3::new(dx, dy, 0.0) * pan_factor;
+        let camera_rotation =
+            Rotation3::new(Vector3::z() * (self.azimuthal_angle - f32::consts::FRAC_PI_2));
+        self.origin += camera_rotation * ground_translation;
+    }
+
+    pub fn pan_screen(&mut self, dx: f32, dy: f32) {
+        let pan_factor = self.options.speed_pan * self.radius / self.options.radius_max;
+        let ground_translation = Vector3::new(dx, dy, 0.0) * pan_factor;
+        let camera_rotation =
+            Rotation3::new(Vector3::z() * (self.azimuthal_angle - f32::consts::FRAC_PI_2));
+
+        // Compute normal vector of the screen plane
+        let eye = self.compute_eye();
+        let normal = eye - self.origin;
+
+        // Create rotation from XY plane to screen plane
+        // Note that the vectors can theoretically be zero... just don't do anything in that case.
+        if let Some(xy_to_screen_rotation) = Rotation3::rotation_between(&self.up, &normal) {
+            self.origin += xy_to_screen_rotation * camera_rotation * ground_translation;
+        }
+    }
+
+    pub fn rotate(&mut self, dtheta: f32, dphi: f32) {
+        let dtheta = dtheta * self.options.speed_rotate;
+        let dphi = dphi * self.options.speed_rotate;
+
+        self.azimuthal_angle = fmod(self.azimuthal_angle + dtheta, TWO_PI);
+        self.polar_angle = clamp(
+            self.polar_angle + dphi,
+            self.options.polar_angle_distance_min,
+            f32::consts::PI - self.options.polar_angle_distance_min,
+        );
+    }
+
+    pub fn zoom(&mut self, zoom_scale: f32) {
+        let zoom_speed = ZOOM_SPEED_BASE.powf(self.options.speed_zoom * zoom_scale.abs());
+        let new_radius = match zoom_scale.partial_cmp(&0.0) {
+            Some(Ordering::Greater) => self.radius * zoom_speed,
+            Some(Ordering::Less) => self.radius / zoom_speed,
+            _ => self.radius,
+        };
+
+        self.radius = clamp(new_radius, self.options.radius_min, self.options.radius_max);
+    }
+
+    pub fn zoom_step(&mut self, zoom_steps: i32) {
+        let zoom_speed = ZOOM_SPEED_BASE.powf(self.options.speed_zoom_step);
+
+        let mut new_radius = self.radius;
+        match zoom_steps.cmp(&0) {
```

Comment:
Strangely enough, I find `match` more readable than `if/else` (except maybe for a single `if` or `if let`), but I can change it if we agree otherwise.